aten/src/ATen/SparseGCSTensorImpl.cpp

```diff
@@ -0,0 +1,130 @@
+      , at::empty({0}, at::initialTensorOptions().device(sparseGCSTensorSetToDeviceType(key_set)).dtype(ScalarType::Long))
+      , at::empty({0}, at::initialTensorOptions().device(sparseGCSTensorSetToDeviceType(key_set)).dtype(data_type))
+      , at::empty({0}, at::initialTensorOptions().device(sparseGCSTensorSetToDeviceType(key_set)).dtype(ScalarType::Long))
+      , Scalar()  ) {}
+
+SparseGCSTensorImpl::SparseGCSTensorImpl(at::DispatchKeySet key_set,
+                                         const caffe2::TypeMeta& data_type,
+                                         at::Tensor pointers, at::Tensor indices, at::Tensor values,
+                                         at::Tensor reduction, Scalar fill_value)
+  : TensorImpl(key_set, data_type, values.device()),
+    pointers_(std::move(pointers)),
+    indices_(std::move(indices)),
+    values_(std::move(values)),
+    reduction_(std::move(reduction)),
+    fill_value_(std::move(fill_value)) {}
+
+void SparseGCSTensorImpl::resize_(IntArrayRef size) {
+    
+}
+
+void SparseGCSTensorImpl::resize_and_clear_(int64_t nnz_size, int64_t ptr_size, int64_t redux_size, IntArrayRef size) {
+  // TODO: perform error checking.
+
+  // call pointers().options() here since the struct contructor calls the tensor constructor
+  // with args for device specific init.
+  auto empty_pointers = at::empty(ptr_size, pointers().options());
+  auto empty_indices = at::empty(nnz_size, indices().options());
+  auto empty_values = at::empty(nnz_size, values().options());
+  auto empty_reduction = at::empty(redux_size, reduction().options());
+
+  // directly set to the member variables. there should be lots of error checking here.
+  pointers_ = empty_pointers;
+  indices_ = empty_indices;
+  values_ = empty_values;
+  reduction_ = empty_reduction;
+  sizes_ = size.vec();
+}
+
+void SparseGCSTensorImpl::resize_as_(const Tensor& src) {
+  pointers_ = at::empty_like(src.pointers(), src.pointers().options(), src.pointers().suggest_memory_format());
+  indices_ = at::empty_like(src.indices(), src.indices().options(), src.indices().suggest_memory_format());
+  values_ = at::empty_like(src.values(), src.values().options(), src.values().suggest_memory_format());
+  reduction_ = at::empty_like(src.reduction(), src.reduction().options(), src.reduction().suggest_memory_format());
+  sizes_ = src.sizes();
+}
+  
+void SparseGCSTensorImpl::set_member_tensors_unsafe(const Tensor& pointers, const Tensor& indices,
+                                                      const Tensor& values, const Tensor& reduction,
+                                                    const Scalar& fill_value) {
+  // TODO: perform lots of error checking to check correct type and sizes of inputs. Check
+  // SparseTensorImpl::set_indices_and_values_unsafe() for details
+  TORCH_CHECK(!indices.is_sparse(), "expected indices to be a dense tensor, but got indices of layout ", indices.layout());
```

Comment:
How about changing the present `is_sparse()` to `is_sparse(COO_LAYOUT)`? It can accept an enum that with a default set to COO to check what the layout is.

Reply:
Don't forget about BC concerns. No matter what you do, `is_sparse()` has to be kept reporting COO sparsity until you can do a deprecation cycle around it. I don't see what the problem with having `is_coo_sparse` and `is_gcs_sparse` methods is, no need to make things complicated.

Reply:
So I've kept the `is_sparse` the way it is and added a new function `is_sparse_gcs` that will work only with GCS tensors. Once we have changed the namespace of the COO tensor we can change the `is_sparse` to `is_sparse_coo`.